File: kotlin/src/main/kotlin/com/bhengubv/circleai/plugins/Plugins.kt
// Plugins.kt
//
// Kotlin port of CircleAI.Plugins — the C# reference is the EXACT spec for the
// plugin contract surface (IPlugin.cs, PluginContext.cs). Ports the stable
// surface a plugin is allowed to use: the plugin contract, the plugin context,
// the string-keyed event bus, and the permission-gated context wrapper.
//
// The C# assembly-loading / hot-reload machinery (PluginLoader,
// PluginRegistry, PluginLifecycleService) is reflection + IHostedService +
// AssemblyLoadContext specific and is intentionally out of scope for the
// deterministic in-memory Kotlin core.
//
// C# -> Kotlin conventions:
//   Task                 -> suspend fun
//   IDisposable          -> AutoCloseable
//   Action<object?>      -> (Any?) -> Unit
//   ILogger              -> PluginLogger (minimal local abstraction)
//   ConcurrentDictionary -> synchronized MutableMap

package com.bhengubv.circleai.plugins

// ===========================================================================
// Logging abstraction
// ===========================================================================

/**
 * Minimal logger a plugin receives through its context. Stands in for the C#
 * `Microsoft.Extensions.Logging.ILogger`; hosts provide a real implementation.
 */
fun interface PluginLogger {
    fun log(level: PluginLogLevel, message: String, error: Throwable?)
}

enum class PluginLogLevel { Trace, Debug, Information, Warning, Error, Critical }

/** A logger that drops everything on the floor. Default when the host wires none. */
object NullPluginLogger : PluginLogger {
    override fun log(level: PluginLogLevel, message: String, error: Throwable?) {}

    /** Convenience overload for message-only logging. */
    fun log(level: PluginLogLevel, message: String) = log(level, message, null)
}

// ===========================================================================
// IPlugin  (IPlugin.cs)
// ===========================================================================

/**
 * Contract every CircleAI plugin implements. Plugins subscribe to host events
 * through [IPluginContext.events] and read the configured workspace path
 * through [IPluginContext.workspacePath].
 */
interface IPlugin {
    /** Unique identifier (matches the assembly name by default). */
    val id: String

    /** Human-readable label. */
    val displayName: String

    /** SemVer string. */
    val version: String

    /** Called once at host startup. */
    suspend fun initialize(context: IPluginContext)

    /** Called when the host is shutting down or the plugin is being unloaded. */
    suspend fun shutdown()
}

/**
 * Stable surface plugins are allowed to use. Deliberately does not expose a
 * service collection — plugins should not swap out host services. They get an
 * event bus, a logger, and the configured workspace path.
 */
interface IPluginContext {
    /** Host-configured workspace directory (or null when not set). */
    val workspacePath: String?

    /** Event bus the host raises events into. */
    val events: IPluginEvents

    /** Logger scoped to this plugin. */
    val logger: PluginLogger
}

/**
 * String-keyed event bus. The host raises events via [raise]; plugins
 * subscribe with [subscribe]. Payload is opaque; senders + listeners agree on
 * the concrete type per event name.
 */
interface IPluginEvents {
    /** Subscribe to events. Returns an unsubscribe handle. */
    fun subscribe(eventName: String, handler: (Any?) -> Unit): AutoCloseable

    /** Raise an event. Host-only API. */
    fun raise(eventName: String, payload: Any?)
}

/** Thread-safe default [IPluginEvents]. */
class PluginEvents : IPluginEvents {
    private val handlers = HashMap<String, MutableList<(Any?) -> Unit>>()
    private val lock = Any()

    override fun subscribe(eventName: String, handler: (Any?) -> Unit): AutoCloseable {
        require(eventName.isNotEmpty()) { "eventName required" }
        synchronized(lock) { handlers.getOrPut(eventName) { ArrayList() }.add(handler) }
        return Subscription(eventName, handler)
    }

    override fun raise(eventName: String, payload: Any?) {
        val snapshot = synchronized(lock) { handlers[eventName]?.toList() } ?: return
        for (h in snapshot) {
            try {
                h(payload)
            } catch (ex: Throwable) {
                // an unhealthy plugin must not corrupt the host
            }
        }
    }

    private fun unsubscribe(eventName: String, handler: (Any?) -> Unit) {
        synchronized(lock) { handlers[eventName]?.remove(handler) }
    }

    private inner class Subscription(
        private val name: String,
        private val handler: (Any?) -> Unit,
    ) : AutoCloseable {
        private var disposed = false
        override fun close() {
            if (disposed) return
            disposed = true
            unsubscribe(name, handler)
        }
    }
}

/** Well-known event names hosts raise and plugins subscribe to. */
object PluginEventNames {
    const val WORKSPACE_LOADED = "workspace.loaded"
    const val CHAT_MESSAGE = "chat.message"
    const val MODEL_LOADED = "model.loaded"
    const val MODEL_UNLOADED = "model.unloaded"
}

// ===========================================================================
// PluginContext / PermissionedPluginContext  (PluginContext.cs)
// ===========================================================================

/** Default [IPluginContext]. */
class PluginContext(
    private val workspacePathAccessor: () -> String?,
    override val events: IPluginEvents,
    override val logger: PluginLogger,
) : IPluginContext {
    override val workspacePath: String?
        get() = workspacePathAccessor()
}

/**
 * Wraps an inner context and gates capabilities by a granted-permission set.
 * Permission-denied plugins see a null workspace path and a silent event bus.
 */
class PermissionedPluginContext(
    private val inner: IPluginContext,
    grantedPermissions: Iterable<String>,
) : IPluginContext {

    object Permissions {
        const val WORKSPACE_READ = "workspace.read"
        const val WORKSPACE_WRITE = "workspace.write"
        const val EVENTS_SUBSCRIBE = "events.subscribe"
    }

    private val granted: Set<String> = grantedPermissions.map { it.lowercase() }.toHashSet()
    private val gatedEvents: IPluginEvents =
        if (granted.contains(Permissions.EVENTS_SUBSCRIBE)) inner.events else SilentEvents

    override val workspacePath: String?
        get() = if (granted.contains(Permissions.WORKSPACE_READ) || granted.contains(Permissions.WORKSPACE_WRITE)) {
            inner.workspacePath
        } else {
            null
        }

    override val events: IPluginEvents get() = gatedEvents
    override val logger: PluginLogger get() = inner.logger

    /** Drop-on-the-floor event bus for permission-denied plugins. */
    private object SilentEvents : IPluginEvents {
        override fun subscribe(eventName: String, handler: (Any?) -> Unit): AutoCloseable = AutoCloseable { }
        override fun raise(eventName: String, payload: Any?) {}
    }
}

File: swift/Sources/CircleAI/Plugins.swift
// Plugins.swift
//
// Port of the named CircleAI.Plugins/ types — the stable plugin contract
// surface + a string-keyed event bus + a permission-gated context wrapper.
//   • IPlugin.cs         — IPlugin, IPluginContext, IPluginEvents, PluginEvents,
//                          PluginEventNames
//   • PluginContext.cs   — PluginContext (default), PermissionedPluginContext
//   • PluginRegistry.cs  — RegisteredPlugin, PluginRegistry (in-memory port)
//
// Porting notes:
//   • `Microsoft.Extensions.Logging.ILogger` → the tree's `ICircleAILogger`
//     (from Auditing.swift) — the SDK's logging seam.
//   • `IDisposable Subscribe(string, Action<object?>)` → `subscribe(_:_:)
//     -> IPluginSubscription` (idempotent dispose).
//   • The event payload (`object?`) → `(any Sendable)?` so it can cross the
//     handler boundary safely; senders + listeners agree on the concrete type
//     per event name (the C# contract).
//   • Assembly-reflection loading (`PluginLoader`), the DI hosted service
//     (`PluginLifecycleService`), and the JSON-file marketplace are host /
//     runtime infrastructure with no portable Swift analogue and are out of the
//     named work-unit scope. `PluginRegistry` is ported as a pure in-memory
//     registry (its C# JSON persistence is dropped; the register / enable /
//     grant / revoke / uninstall semantics are preserved).

import Foundation

// MARK: - Subscription handle

/// A disposable event subscription. Mirrors the C# `IDisposable` returned by
/// `IPluginEvents.Subscribe`. `dispose()` is idempotent.
public protocol IPluginSubscription: AnyObject, Sendable {
    /// Unsubscribe. Idempotent.
    func dispose()
}

// MARK: - IPluginEvents

/// String-keyed event bus. The host raises events via `raise`; plugins
/// subscribe with `subscribe`. Payload is opaque; senders + listeners agree on
/// the concrete type per event name. (C# `IPluginEvents`.)
public protocol IPluginEvents: Sendable {
    /// Subscribe to `eventName`. Returns a dispose handle.
    func subscribe(_ eventName: String, _ handler: @escaping @Sendable ((any Sendable)?) -> Void) -> IPluginSubscription
    /// Raise `eventName` with `payload`. Host-only API.
    func raise(_ eventName: String, _ payload: (any Sendable)?)
}

/// Thread-safe default `IPluginEvents`. Fan-out is snapshot-then-release so a
/// handler that (un)subscribes cannot self-deadlock, and an unhealthy handler
/// cannot corrupt the host (throws are swallowed — Swift closures can't throw
/// here, so this is inherent). (C# `PluginEvents`.)
public final class PluginEvents: IPluginEvents, @unchecked Sendable {
    private let lock = NSLock()
    private var handlers: [String: [UUID: @Sendable ((any Sendable)?) -> Void]] = [:]

    public init() {}

    public func subscribe(_ eventName: String, _ handler: @escaping @Sendable ((any Sendable)?) -> Void) -> IPluginSubscription {
        precondition(!eventName.isEmpty, "eventName required")
        let id = UUID()
        lock.lock(); handlers[eventName, default: [:]][id] = handler; lock.unlock()
        return Subscription(owner: self, name: eventName, id: id)
    }

    public func raise(_ eventName: String, _ payload: (any Sendable)?) {
        lock.lock()
        let snapshot = handlers[eventName].map { Array($0.values) } ?? []
        lock.unlock()
        for h in snapshot { h(payload) }
    }

    /// Subscriber count for `eventName`. Useful in tests.
    public func subscriberCount(_ eventName: String) -> Int {
        lock.lock(); defer { lock.unlock() }
        return handlers[eventName]?.count ?? 0
    }

    private func unsubscribe(_ eventName: String, _ id: UUID) {
        lock.lock(); handlers[eventName]?[id] = nil; lock.unlock()
    }

    private final class Subscription: IPluginSubscription, @unchecked Sendable {
        private weak var owner: PluginEvents?
        private let name: String
        private let id: UUID
        private let disposeLock = NSLock()
        private var disposed = false

        init(owner: PluginEvents, name: String, id: UUID) {
            self.owner = owner; self.name = name; self.id = id
        }

        func dispose() {
            disposeLock.lock()
            if disposed { disposeLock.unlock(); return }
            disposed = true
            disposeLock.unlock()
            owner?.unsubscribe(name, id)
        }
    }
}

/// Well-known event names. (C# `PluginEventNames`.)
public enum PluginEventNames {
    public static let workspaceLoaded = "workspace.loaded"
    public static let chatMessage = "chat.message"
    public static let modelLoaded = "model.loaded"
    public static let modelUnloaded = "model.unloaded"
}

// MARK: - IPluginContext

/// Stable surface plugins may use: an event bus, a logger, and the configured
/// workspace path. Deliberately does NOT expose service registration.
/// (C# `IPluginContext`.)
public protocol IPluginContext: Sendable {
    /// Host-configured workspace directory (or `nil`).
    var workspacePath: String? { get }
    /// Event bus the host raises events into.
    var events: any IPluginEvents { get }
    /// Logger scoped to this plugin.
    var logger: any ICircleAILogger { get }
}

/// Default `IPluginContext`. The workspace path is resolved lazily via an
/// accessor closure (C# `Func<string?>`). (C# `PluginContext`.)
public final class PluginContext: IPluginContext, @unchecked Sendable {
    private let workspacePathAccessor: @Sendable () -> String?

    public init(workspacePathAccessor: @escaping @Sendable () -> String?,
                events: any IPluginEvents, logger: any ICircleAILogger) {
        self.workspacePathAccessor = workspacePathAccessor
        self.events = events
        self.logger = logger
    }

    public var workspacePath: String? { workspacePathAccessor() }
    public let events: any IPluginEvents
    public let logger: any ICircleAILogger
}

/// Wraps an inner context and gates capabilities by a granted-permission set.
/// (C# `PermissionedPluginContext`.)
public final class PermissionedPluginContext: IPluginContext, @unchecked Sendable {
    /// Well-known permission strings. (C# nested `Permissions`.)
    public enum Permissions {
        public static let workspaceRead = "workspace.read"
        public static let workspaceWrite = "workspace.write"
        public static let eventsSubscribe = "events.subscribe"
    }

    private let inner: any IPluginContext
    private let granted: Set<String>
    private let gatedEvents: any IPluginEvents

    public init(inner: any IPluginContext, grantedPermissions: [String]) {
        self.inner = inner
        // Case-insensitive membership (matches the C# OrdinalIgnoreCase set).
        self.granted = Set(grantedPermissions.map { $0.lowercased() })
        self.gatedEvents = self.granted.contains(Permissions.eventsSubscribe.lowercased())
            ? inner.events
            : SilentEvents()
    }

    public var workspacePath: String? {
        if granted.contains(Permissions.workspaceRead.lowercased())
            || granted.contains(Permissions.workspaceWrite.lowercased()) {
            return inner.workspacePath
        }
        return nil
    }

    public var events: any IPluginEvents { gatedEvents }
    public var logger: any ICircleAILogger { inner.logger }

    /// Drop-on-the-floor event bus for permission-denied plugins.
    private final class SilentEvents: IPluginEvents, @unchecked Sendable {
        func subscribe(_ eventName: String, _ handler: @escaping @Sendable ((any Sendable)?) -> Void) -> IPluginSubscription {
            NoopSubscription.shared
        }
        func raise(_ eventName: String, _ payload: (any Sendable)?) {}
    }

    private final class NoopSubscription: IPluginSubscription, @unchecked Sendable {
        static let shared = NoopSubscription()
        func dispose() {}
    }
}

// MARK: - IPlugin

/// Contract every CircleAI plugin implements. (C# `IPlugin`.)
public protocol IPlugin: Sendable {
    /// Unique identifier.
    var id: String { get }
    /// Human-readable label.
    var displayName: String { get }
    /// SemVer string.
    var version: String { get }
    /// Called once at host startup with the plugin context.
    func initialize(context: any IPluginContext) async throws
    /// Called on host shutdown / plugin unload.
    func shutdown() async throws
}

// MARK: - PluginRegistry (in-memory port)

/// One installed-plugin entry. (C# `RegisteredPlugin`.)
public struct RegisteredPlugin: Sendable, Equatable, Codable {
    public var id: String
    public var displayName: String
    public var version: String
    public var permissions: [String]
    public var enabled: Bool
    public var installedAt: Date

    public init(id: String, displayName: String, version: String, permissions: [String],
                enabled: Bool, installedAt: Date) {
        self.id = id
        self.displayName = displayName
        self.version = version
        self.permissions = permissions
        self.enabled = enabled
        self.installedAt = installedAt
    }
}

/// Tracks installed plugins. Declarative, opt-in permissions per plugin. The C#
/// version persists to a JSON manifest; this in-memory port keeps the same
/// register / enable / grant / revoke / uninstall semantics without file I/O.
/// (C# `PluginRegistry`.)
public final class PluginRegistry: @unchecked Sendable {
    private let lock = NSLock()
    private var installed: [RegisteredPlugin] = []

    public init() {}

    /// All installed entries (snapshot).
    public var all: [RegisteredPlugin] {
        lock.lock(); defer { lock.unlock() }
        return installed
    }

    /// Returns the entry with `id` (case-insensitive), or `nil`.
    public func get(_ id: String) -> RegisteredPlugin? {
        lock.lock(); defer { lock.unlock() }
        return installed.first { $0.id.caseInsensitiveCompare(id) == .orderedSame }
    }

    /// Registers (or replaces, by id) a plugin. New entries start disabled.
    @discardableResult
    public func register(id: String, displayName: String, version: String, permissions: [String]) -> RegisteredPlugin {
        let entry = RegisteredPlugin(id: id, displayName: displayName, version: version,
                                     permissions: permissions, enabled: false, installedAt: Date())
        lock.lock()
        installed.removeAll { $0.id.caseInsensitiveCompare(id) == .orderedSame }
        installed.append(entry)
        lock.unlock()
        return entry
    }

    /// Enables/disables a plugin. Returns false when unknown.
    @discardableResult
    public func setEnabled(_ id: String, _ enabled: Bool) -> Bool {
        lock.lock(); defer { lock.unlock() }
        guard let idx = installed.firstIndex(where: { $0.id.caseInsensitiveCompare(id) == .orderedSame }) else { return false }
        installed[idx].enabled = enabled
        return true
    }

    /// Grants a permission (idempotent). Returns false when the plugin is unknown.
    @discardableResult
    public func grantPermission(_ id: String, _ permission: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        guard let idx = installed.firstIndex(where: { $0.id.caseInsensitiveCompare(id) == .orderedSame }) else { return false }
        if !installed[idx].permissions.contains(where: { $0.caseInsensitiveCompare(permission) == .orderedSame }) {
            installed[idx].permissions.append(permission)
        }
        return true
    }

    /// Revokes a permission. Returns true when something was removed.
    @discardableResult
    public func revokePermission(_ id: String, _ permission: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        guard let idx = installed.firstIndex(where: { $0.id.caseInsensitiveCompare(id) == .orderedSame }) else { return false }
        let before = installed[idx].permissions.count
        installed[idx].permissions.removeAll { $0.caseInsensitiveCompare(permission) == .orderedSame }
        return installed[idx].permissions.count != before
    }

    /// Uninstalls a plugin. Returns true when an entry was removed.
    @discardableResult
    public func uninstall(_ id: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        let before = installed.count
        installed.removeAll { $0.id.caseInsensitiveCompare(id) == .orderedSame }
        return installed.count != before
    }
}
